R/derive_vars_dy.R

```diff
@@ -0,0 +1,87 @@
+#' Derive Analysis Relative Day
+#'
+#' Adds the analysis relative day (`--DY`) to the dataset, i.e., study
+#' day of analysis date.
+#'
+#' @param dataset Input dataset
+#'
+#'   The columns specified by the `reference_date` and the `source_vars` parameter are
+#'   expected.
+#'
+#' @param reference_date The start date column, e.g., date of first treatment
+#'
+#'   A date or date-time object column is expected.
+#'
+#' @param source_vars A list of datetime or date variables created using `vars()` from
+#'   which dates are to be extracted.
+#'
+#' @author Teckla Akinyi
+#'
+#' @details The study day is derived as number of days from the reference date
+#'   to the end date. If it is nonnegative, one is added. I.e., the study day of the
+#'   reference date is 1.
+#'
+#' @return The input dataset with `--DY` corresponding to the `--DTM` or `--DT`
+#'     source variable(s) added
+#'
+#' @keywords derivation bds occds timing
+#'
+#' @export
+#'
+#' @examples
+#' datain <- tibble::tribble(
+#'  ~STUDYID, ~USUBJID, ~TRTSDTM, ~ASTDTM, ~AENDT,
+#'  "TEST01", "PAT01", "2014-01-17T23:59:59", "2014-01-18T13:09:O9", "2014-01-20"
+#' ) %>%
+#'  mutate(TRTSDTM=lubridate::as_datetime(TRTSDTM),
+#'         ASTDT=lubridate::as_datetime(ASTDTM),
+#'         AENDT=lubridate::ymd(AENDT))
+#'
+#' derive_vars_dy(datain, reference_date = TRTSDTM,source_vars = vars(TRTSDTM, ASTDTM, AENDT))
+#'
+#'  datain <- tibble::tribble(
+#'  ~STUDYID, ~USUBJID, ~TRTSDT, ~ASTDTM, ~AENDT,
+#'  "TEST01", "PAT01", "2014-01-17", "2014-01-18T13:09:O9", "2014-01-20"
+#'  ) %>%
+#'  mutate(TRTSDTM=lubridate::ymd(TRTSDT),
+#'         ASTDT=lubridate::as_datetime(ASTDTM),
+#'         AENDT=lubridate::ymd(AENDT))
+#'
+#'  derive_vars_dy(datain, reference_date = TRTSDT, source_vars = vars(TRTSDT, ASTDTM, AENDT))
+derive_vars_dy <- function(dataset,
+                           reference_date,
+                           source_vars) {
+  #assertions
+  reference_date <- assert_symbol(enquo(reference_date))
+  assert_data_frame(dataset, required_vars = quo_c(source_vars, reference_date))
+
+  #Warn if `--DY` variables already exist
+  dtm_vars <- vars2chr(source_vars)
+  n_vars <- length(source_vars)
+
+  dy_vars <- sub("DTM$", "DY", sub("DT$", "DY", dtm_vars))
+  warn_if_vars_exist(dataset, dy_vars)
+
+  #call date part of reference date if --DTM is parsed
+
+  if (n_vars > 1L) {
+    dataset %>%
+      mutate_at(source_vars,
+                .funs = list(temp = ~ ceiling(as.numeric(
+                                              difftime(., `!!`(reference_date), unit = "days")))
+                             )
+                ) %>%
+      # if ge to 0 then add 1
+      mutate_at(vars(ends_with("temp")), ~ ifelse(.x >= 0, .x + 1, .x)) %>%
```

Comment:
I wonder if it makes sense to put this into a little helper function, something like:

```r
compute_dy <- function(date, reference_date) {
  day_diff <- ceiling(as.numeric(difftime(date, reference_date, unit = "days")))
  day_diff + if_else(day_diff >= 0, 1, 0)
}
```

That way you don't replicate the code in the `if` and `else` branch.

Reply:
Why should we reimplement the study day computation?
What about the following?
```
compute_dy <- function(date, reference_date) {
  compute_duration(
    start_date = reference_date,
    end_date = date)
}
```

Reply:
on the comment from @thomas-neitmann  - I would have to think more on this since I use mutate_at(source_vars) and mutate() for a single var....I may have to brute force mutate_at()  to handle the case where one dt(m) is parsed and it ends up being a bit circular relative to the current implementation.
@bundfussr I thought the whole idea of having this function is to expose a direct derivation of day sans dependencies otherwise I agree, what is the point might as well directly call compute_duration and there is absolutely no need for a wrapper, then compute_duration can simply be expanded to handle multiple dates.

Reply:
I thought the intention is to improve the user interface as users looking for a study day derivations may miss that it can be done by `derive_duration()`.